### DISCUSSION TITLE
HA- Confguration Using EC2 instances and NLBs

### DISCUSSION THREAD

**Discussion by Sudhigalagali** [StackStorm/st2#5586] [Operations]
Need help!...

We are trying out ST2 in HA. We do not have prior knowledge of ST2. We are trying out this for the first time. 
If our POC successful we will be implementing this in our automation tasks.

2 nodes - WebGUI
2 Nodes - ST2 Components ( Blue Prints)
NLB 1  -    For WebGUI
NLB 2  -    For communication between WebGUI and ST2 Blue print nodes)
RabbitMQ - AWS Managed Service
Mongo DB - EC2 Instances Clustered

Operating System is Linux. ST2 version 3.6

What I want to know is,

1. Will this be a recommended setup? 
2. How to make sure that ST blue print nodes writes/ access data from same location? Shouldn't they be using using some central filesystem? 

This might be a silly query but I need some guidance here.  Attached file will give will help you in visualizing setup.

Thanks much
Sudhi

![HA](https://user-images.githubusercontent.com/84116857/156619126-cd9553c3-59e1-4db5-a61a-bc56d90f8c76.JPG)
i

**Comment by arm4b** [comment]
The approach looks good overall :+1: 

Take a look at
https://docs.stackstorm.com/reference/ha.html which highlights some edge cases about each of st2 microservices and recommends the shared file system for pack content like NFS.

Additionally, there is https://github.com/StackStorm/stackstorm-ha/ Helm chart with K8s which codifies the HA deployment in various ways.

I'd recommend taking a look at those as food for thought.

Alternatively, if you have resources you can consult with the StackStorm Partners: https://stackstorm.com/partners/ who may guide you in deeper details and solutions.

**Reply by Sudhigalagali** [reply]
Thank you @armab 

I think you are referring to below for NFS...

************************************************************************************************
Read-Write NFS mounts
If the content folders i.e. /opt/stackstorm/packs and /opt/stackstorm/virtualenvs are placed on read-write NFS mounts then writes from any StackStorm node will be visible to other nodes. Special care needs to be taken with /opt/stackstorm/virtualenvs since that has symlinks to system libraries. If care is not taken to provision all host boxes in an identical manner it could lead to unpredictable behavior. Managing the virtualenvs on every host box individually would be a more robust approach
************************************************************************************************

So if I understand this correctly I need to mount */packs/ and */virtualenvs folders to a NFS filesystem. on both the blueprint nodes. 
Currently I see that */packs folder has some more folders. Those will be recreated on the NFS filesystem ? or will have to be copied over? */virtualenvs/ is empty... 

ST2 is installed/ configured as per diagram and no workflow has been created yet.

Thanks,
Sudhi
